executor/common_linux.h

```diff
@@ -2509,50 +2544,42 @@ static long syz_read_part_table(volatile unsigned long size, volatile unsigned l
 // }
 static long syz_mount_image(volatile long fsarg, volatile long dir, volatile unsigned long size, volatile unsigned long nsegs, volatile long segments, volatile long flags, volatile long optsarg)
 {
-	char loopname[64], fs[32], opts[256];
-	int loopfd, err = 0, res = -1;
-	unsigned long i;
-
-	size = fs_image_segment_check(size, nsegs, segments);
-	int memfd = syscall(sys_memfd_create, "syz_mount_image", 0);
-	if (memfd == -1) {
-		err = errno;
-		goto error;
-	}
-	if (ftruncate(memfd, size)) {
-		err = errno;
-		goto error_close_memfd;
-	}
-	for (i = 0; i < nsegs; i++) {
-		struct fs_image_segment* segs = (struct fs_image_segment*)segments;
-		if (pwrite(memfd, segs[i].data, segs[i].size, segs[i].offset) < 0) {
-			debug("syz_mount_image: pwrite[%u] failed: %d\n", (int)i, errno);
+	char loopname[64], opts[256];
+	char* source = NULL;
+	struct fs_image_segment* segs;
+	int need_loop_device;
+	int res = -1;
+	int err = 0;
+	int loopfd;
+	int memfd;
+	char* mount_opts;
+	char* target;
+	char* fs;
+
+	fs = (char*)fsarg;
+	target = (char*)dir;
+	segs = (struct fs_image_segment*)segments;
+	mount_opts = (char*)optsarg;
+	need_loop_device = !!segs;
+
+	if (need_loop_device) {
+		// Some filesystems (e.g. FUSE) do not need a backing device or
+		// filesystem image.
+		memset(loopname, 0, sizeof(loopname));
+		snprintf(loopname, sizeof(loopname), "/dev/loop%llu", procid);
+		if (setup_loop_device(size, nsegs, segs, "syz_mount_image", loopname, &memfd, &loopfd) == -1) {
+			return -1;
 		}
+		source = loopname;
 	}
-	snprintf(loopname, sizeof(loopname), "/dev/loop%llu", procid);
-	loopfd = open(loopname, O_RDWR);
-	if (loopfd == -1) {
-		err = errno;
-		goto error_close_memfd;
-	}
-	if (ioctl(loopfd, LOOP_SET_FD, memfd)) {
-		if (errno != EBUSY) {
-			err = errno;
-			goto error_close_loop;
-		}
-		ioctl(loopfd, LOOP_CLR_FD, 0);
-		usleep(1000);
-		if (ioctl(loopfd, LOOP_SET_FD, memfd)) {
-			err = errno;
-			goto error_close_loop;
-		}
+
+	// Leave some space for the additional options we append below.
+	if (strlen(mount_opts) > (sizeof(opts) - 32)) {
+		debug("ERROR: syz_mount_image parameter optsarg bigger than internal opts\n");
 	}
-	mkdir((char*)dir, 0777);
-	memset(fs, 0, sizeof(fs));
-	strncpy(fs, (char*)fsarg, sizeof(fs) - 1);
+	mkdir(target, 0777);
 	memset(opts, 0, sizeof(opts));
-	// Leave some space for the additional options we append below.
```

Comment:
Restore the comment.

Reply:
The comment is still there, I have just moved it up a little where we check if are truncating the received mount options. I don't think this will ever happen but having a debug in case that happens sounds ok to me, what do you think?